### PR TITLE
Add an endpoint returning a national or a local poll

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Controller/Api/PollController.php [hunk 1]
```diff
@@ -67,4 +70,28 @@ public function vote(
...
+        if (!$poll && $region = $zoneRepository->findRegionByPostalCode($postalCode)) {
+            $poll = $localPollRepository->findOnePublishedByZone($region);
+        }
+
+        return new JsonResponse(
+            $serializer->serialize($poll, 'json', ['groups' => ['poll_read']]),
+            JsonResponse::HTTP_OK,
+            [],
+            true
+        );
```
Comment (ottaviano) [inline src/Controller/Api/PollController.php:95]:
```suggestion
        return $this->json($poll, Response::HTTP_OK, [], ['groups' => ['poll_read']]);
```

File: src/Entity/Poll/Poll.php [hunk 2]
```diff
@@ -173,8 +173,10 @@ public function getResult(): array
             ];
         }
 
-        foreach ($result['choices'] as $id => $choice) {
-            $result['choices'][$id]['percentage'] = round($choice['count'] / $result['total'] * 100, 1);
+        if (0 !== $result['total']) {
+            foreach ($result['choices'] as $id => $choice) {
+                $result['choices'][$id]['percentage'] = round($choice['count'] / $result['total'] * 100, 1);
+            }
```
Comment (ottaviano) [inline src/Entity/Poll/Poll.php:179]:
```suggestion
        foreach ($result['choices'] as $id => $choice) {
            $result['choices'][$id]['percentage'] = $result['total'] > 0 ? round($choice['count'] / $result['total'] * 100, 1) : 0;
```

File: src/Normalizer/PollNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+use App\Entity\Poll\LocalPoll;
+use App\Entity\Poll\NationalPoll;
+use App\Entity\Poll\Poll;
+use App\Entity\Poll\PollTypeEnum;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareTrait;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+class PollNormalizer implements NormalizerInterface, NormalizerAwareInterface
```
Comment (ottaviano) [inline src/Normalizer/PollNormalizer.php:13]: we can replace this Normalizer by Poll class method, for exemple LocalPoll: 
```php
class LocalPoll extends Poll 
{
  // ... 
  /**
   * @SerializerGroup("....")
   */
  public function getType(): string
  {
    // ...
  }
}



```

File: src/Repository/Geo/ZoneRepository.php [hunk 4]
```diff
@@ -298,6 +298,28 @@ public function findByPostalCode(string $postalCode): array
...
+                'postal_code_1' => $postalCode.'%',
+                'postal_code_2' => '%,'.$postalCode.'%',
+                'region' => Zone::REGION,
+                'city' => Zone::CITY,
+            ])
+            ->getQuery()
+            ->getResult()
+        ;
+
+        return 1 === \count($regions) ? $regions[0] : null;
```
Comment (ottaviano) [inline src/Repository/Geo/ZoneRepository.php:320]:
```suggestion
        return $this->createQueryBuilder('zone')
            ->leftJoin('zone.children', 'child')
            ->innerJoin(City::class, 'city', Join::WITH, 'child.code = city.code')
            ->where('(city.postalCode LIKE :postal_code_1 OR city.postalCode LIKE :postal_code_2)')
            ->andWhere('zone.type = :region AND child.type = :city')
            ->setParameters([
                'postal_code_1' => $postalCode.'%',
                'postal_code_2' => '%,'.$postalCode.'%',
                'region' => Zone::REGION,
                'city' => Zone::CITY,
            ])
            ->getQuery()
            ->setMaxResults(1)
            ->getOneOrNullResult()
        ;
```

File: src/Repository/Poll/LocalPollRepository.php [hunk 5]
```diff
@@ -44,4 +45,24 @@ public function findAllByZonesWithStats(array $zones): array
...
         ;
     }
+
+    public function findOnePublishedByZone(Zone $zone): ?LocalPoll
+    {
+        return $this
+            ->createQueryBuilder('poll')
+            ->innerJoin('poll.zone', 'zone')
+            ->innerJoin('zone.parents', 'parent')
+            ->innerJoin('zone.children', 'child')
```
Comment (ottaviano) [inline src/Repository/Poll/LocalPollRepository.php:55]: these joins are unused

File: src/Repository/Poll/NationalPollRepository.php [hunk 6]
```diff
@@ -11,4 +12,16 @@ public function __construct(ManagerRegistry $registry)
     {
         parent::__construct($registry, NationalPoll::class);
     }
+
+    public function findLastActivePoll(): ?NationalPoll
+    {
+        return $this->createQueryBuilder('poll')
+            ->where('poll.finishAt > :now')
+            ->orderBy('poll.finishAt', 'desc')
+            ->setParameter('now', new Chronos())
```
Comment (ottaviano) [inline src/Repository/Poll/NationalPollRepository.php:21]: why do you use Chronos here?

Comment (illusionOfParadise) [reply src/Repository/Poll/NationalPollRepository.php:21]: To test the case with no active national poll to return a local poll.

**Review by ottaviano — APPROVED** [review]

**Pull request merged**